### PR TITLE
AP_Networking: added NET_Pn_ network ports for AP_SerialManager

### PR DESCRIPTION
**Pull request by tridge**
This is an alternative approach to allowing UART drivers as AP_Networking ports

Tested on CubeRed and Pixhawk6X, works with MAVLink and gimbal SIYI ZT30

### CODE REVIEW

**Comment by timtuxworth** [comment]
This is brilliant - but is 9 ports really enough? How about NET_nn_

**Comment by magicrub** [comment]
> This is brilliant - but is 9 ports really enough? How about NET_nn_

There's also a whole alphabet after 9 :). That said, there's room (so far) to add double digits later on

File: libraries/AP_Vehicle/AP_Vehicle.cpp [hunk 1]
```diff
@@ -210,7 +210,36 @@ const AP_Param::GroupInfo AP_Vehicle::var_info[] = {
...
+#endif
+
+#if AP_NETWORKING_NUM_PORTS > 2
+    // @Group: NET_P3_
+    // @Path: ../AP_Networking/AP_Networking_port.cpp
+    AP_SUBGROUPINFO(networking.ports[2], "NET_P3_", 24, AP_Vehicle, AP_Networking::Port),
+#endif
+
+#if AP_NETWORKING_NUM_PORTS > 3
+    // @Group: NET_P3_
```
Comment (magicrub) [inline libraries/AP_Vehicle/AP_Vehicle.cpp:238]: NET_P4_

**Comment by timtuxworth** [comment]
> > This is brilliant - but is 9 ports really enough? How about NET_nn_
> 
> There's also a whole alphabet after 9 :). That said, there's room (so far) to add double digits later on

Oh no - I mention it now because you have to catch it now, otherwise you will end up with the same sorting order problem you have now with RC1, RC11, RC12, RC16, RC2, RC3 and SERIAL... and SERVO1, 11, 12 ..  - do you really want the net ports sorted 1, 11, 12, 13 ... 2, 22, 23 .. 3 ?

File: libraries/AP_Networking/AP_Networking.h [hunk 2]
```diff
@@ -175,6 +180,51 @@ class AP_Networking
...
+        }
+        bool tx_pending() override {
+            return false;
+        }
+
+        void udp_client_init(void);
+        void udp_client_loop(void);
+
+    private:
+        bool init_buffers(uint32_t size);
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking.h:213]: Add WARN_IF_UNUSED. If this fails, you need to call it again.

File: libraries/AP_Vehicle/AP_Vehicle.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_Vehicle/AP_Vehicle.cpp]: fixed

File: libraries/AP_Networking/AP_Networking.h (hunk 2 above)
Comment (tridge) [reply libraries/AP_Networking/AP_Networking.h:213]: its a private interface, we normally use WARN_IF_UNUSED on public interfaces

File: libraries/AP_Networking/AP_Networking_port.cpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        }
+    }
+}
+
+/*
+  initialise a UDP client
+ */
+void AP_Networking::Port::udp_client_init(void)
+{
+    if (!init_buffers(4096)) {
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking_port.cpp:78]: This 4096 should be a define

File: libraries/AP_Networking/AP_Networking_port.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        return;
+    }
+    sock = new SocketAPM(true);
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking_port.cpp:81]: should return early, or maybe even throw an error, if sock is non-null

File: libraries/AP_Networking/AP_Networking_port.cpp [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+/*
+  update a UDP client
+ */
+void AP_Networking::Port::udp_client_loop(void)
+{
+    while (!hal.scheduler->is_system_initialized()) {
+        hal.scheduler->delay(100);
+    }
+    hal.scheduler->delay(1000);
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking_port.cpp:98]: is there any better way to do this more intelligently? All over the place in the code we add delays to ensure we don't send GCS msgs out too early... but SURELY there's some sort of status we can check somewhere to know it's a proper time

Comment (magicrub) [reply libraries/AP_Networking/AP_Networking_port.cpp:98]: for what it's worth, on my AP_Periph version of this I had to bump this up a little

File: libraries/AP_Networking/AP_Networking_port.cpp [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            if (ret > 0) {
+                writebuffer->advance(ret);
+            }
+        }
+
+        // handle incoming packets
+        const auto space = readbuffer->space();
+        if (space > 0) {
+            const uint32_t n = MIN(350U, space);
+            uint8_t buf[n];
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking_port.cpp:127]: I'm surprised this passed CLANG. Normally variable length arrays throw warnings

File: libraries/AP_Networking/AP_Networking_port.cpp [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+bool AP_Networking::Port::_discard_input()
+{
+    readbuffer->clear();
+    return true;
+}
+
+/*
+  initialise read/write buffers
+ */
+bool AP_Networking::Port::init_buffers(uint32_t size)
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking_port.cpp:174]: should use different rx and tx buffer sizes

File: libraries/AP_SerialManager/AP_SerialManager.cpp [hunk 8]
```diff
@@ -646,6 +672,31 @@ AP_HAL::UARTDriver *AP_SerialManager::get_serial_by_id(uint8_t id)
     if (id < SERIALMANAGER_NUM_PORTS) {
         return hal.serial(id);
     }
+#if AP_SERIALMANAGER_REGISTER_ENABLED
+    for (auto p = registered_ports; p; p = p->next) {
```
Comment (magicrub) [inline libraries/AP_SerialManager/AP_SerialManager.cpp:676]: we should wrap this in 
`if (id >= AP_SERIALMANAGER_NET_PORT_1 && id <= AP_SERIALMANAGER_NET_PORT_1 + AP_NETWORKING_NUM_PORTS)`

File: libraries/AP_SerialManager/AP_SerialManager.cpp [hunk 9]
```diff
@@ -646,6 +672,31 @@ AP_HAL::UARTDriver *AP_SerialManager::get_serial_by_id(uint8_t id)
...
+/*
+  get a UARTState by index
+*/
+const AP_SerialManager::UARTState *AP_SerialManager::get_state_by_id(uint8_t id) const
+{
+    if (id < SERIALMANAGER_NUM_PORTS) {
+        return &state[id];
+    }
+#if AP_SERIALMANAGER_REGISTER_ENABLED
+    for (auto p = registered_ports; p; p = p->next) {
```
Comment (magicrub) [inline libraries/AP_SerialManager/AP_SerialManager.cpp:694]: needs if `(id >= AP_SERIALMANAGER_NET_PORT_1 && id <= AP_SERIALMANAGER_NET_PORT_1 + AP_NETWORKING_NUM_PORTS)`

File: libraries/AP_Vehicle/AP_Vehicle.cpp [hunk 10]
```diff
@@ -210,7 +210,36 @@ const AP_Param::GroupInfo AP_Vehicle::var_info[] = {
...
+
+    /*
+      the NET_Pn_ parameters need to be in AP_Vehicle as otherwise we
+      are too deep in the parameter tree
+     */
+
+#if AP_NETWORKING_NUM_PORTS > 0
+    // @Group: NET_P1_
+    // @Path: ../AP_Networking/AP_Networking_port.cpp
+    AP_SUBGROUPINFO(networking.ports[0], "NET_P1_", 22, AP_Vehicle, AP_Networking::Port),
```
Comment (magicrub) [inline libraries/AP_Vehicle/AP_Vehicle.cpp:222]: We definitely need to fix this param depth problem in a later PR.

**Review by magicrub — APPROVED** [review]
Approved as-is, all my concerns are minor and shouldn't hold up the merge.

**Comment by magicrub** [comment]
Merging this as-is, I'll make a follow-up PR to address my minor tweaks. Great work fellas!

**Comment by magicrub** [comment]
> > > This is brilliant - but is 9 ports really enough? How about NET_nn_
> > 
> > 
> > There's also a whole alphabet after 9 :). That said, there's room (so far) to add double digits later on
> 
> Oh no - I mention it now because you have to catch it now, otherwise you will end up with the same sorting order problem you have now with RC1, RC11, RC12, RC16, RC2, RC3 and SERIAL... and SERVO1, 11, 12 .. - do you really want the NET ports sorted 1, 11, 12, 13 ... 2, 22, 23 .. 3 ?

@timtuxworth I think 9 ports is plenty. This is just for services using SERIAL_MANAGER. Other things like http web serves and time servers (NTP) won't be consuming "port" numbers here.

**Comment by magicrub** [comment]
follow-up PR: https://github.com/ArduPilot/ardupilot/pull/25572

**Pull request merged**